reconcile/utils/membershipsources/models.py

```diff
@@ -0,0 +1,92 @@
+from collections.abc import Sequence
+from typing import (
+    Any,
+    Callable,
+    Optional,
+    Protocol,
+    TypeVar,
+    Union,
+)
+
+from pydantic import (
+    BaseModel,
+    Extra,
+)
+
+from reconcile.gql_definitions.fragments.membership_source import (
+    MembershipProviderSourceV1,
+    RoleMembershipSource,
+)
+
+
+class User(Protocol):
+    name: str
+    org_username: str
+
+    def dict(self, *, by_alias: bool = False) -> dict[str, Any]:
+        ...
+
+
+class Bot(Protocol):
+    name: str
+    org_username: Optional[str]
+
+    def dict(self, *, by_alias: bool = False) -> dict[str, Any]:
+        ...
+
+
+class RoleWithMemberships(Protocol):
+    name: str
+
+    @property
+    def users(self) -> Sequence[User]:
+        ...
+
+    @property
+    def bots(self) -> Sequence[Bot]:
+        ...
+
+    @property
+    def member_sources(self) -> Optional[Sequence[RoleMembershipSource]]:
+        ...
```

Comment:
the ones i marked as `property` need to be one in order for the Protocol to be safe to use, basically making them read only (see https://mypy.readthedocs.io/en/latest/common_issues.html#covariant-subtyping-of-mutable-protocol-members-is-rejected).

but you are right, it looks strange. i will mark name as property as well because for consistency. i don't mind about it being read only.

Reply:
it is even a good thing for them to be read only. i will change the other protocols as well.